File: dotnet/DesignPatterns/StructuralDesignPatterns/FlyweightDesignPattern/FlyweightDesignPattern.Example1/TreeType.cs
namespace FlyweightDesignPattern.Example1;

/// <summary>
/// Concrete Flyweight class that stores the intrinsic (shared) state.
/// Multiple Tree objects can share the same TreeType instance.
/// This significantly reduces memory usage when rendering a forest with many similar trees.
/// </summary>
public class TreeType : ITreeType
{
    public string Name { get; }
    public string Color { get; }
    public string Texture { get; }

    public TreeType(string name, string color, string texture)
    {
        Name = name;
        Color = color;
        Texture = texture;
    }

    public void Draw(int x, int y)
    {
        Console.WriteLine($"  Drawing '{Name}' tree at ({x}, {y}) - Color: {Color}, Texture: {Texture}");
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/FlyweightDesignPattern/FlyweightDesignPattern.Example1/src/TreeType.ts
import { ITreeType } from './ITreeType';

// Concrete Flyweight — stores intrinsic (shared) state
export class TreeType implements ITreeType {
  constructor(
    public readonly name: string,
    public readonly color: string,
    public readonly texture: string,
  ) {}

  draw(x: number, y: number): void {
    console.log(`  Drawing '${this.name}' tree at (${x}, ${y}) - Color: ${this.color}, Texture: ${this.texture}`);
  }
}
